### ISSUE TITLE
How are batch statistics computed?

### ISSUE THREAD

**Issue by OverLordGoldDragon** [cooijmanstim/recurrent-batch-normalization#6]
I'm implementing recurrent BN in Keras, but looking at the original paper and those citing it, a detail remains unclear to me: how are _batch statistics_ computed? In the original, authors state (pg. 3) (emphasis mine):

> At training time, the statistics E[**h**] and Var[**h**] are estimated by the sample mean and sample variance of the **current minibatch**

Yet [another paper](https://jaywhang.com/assets/batchnorm_rnn.pdf) (pg. 3) using and citing it describes:

> We subscript BN by time (BN_t) to indicate that each time step tracks its own mean and variance. In practice, we **track** these statistics as they change **over the course of training** using an _exponential moving average_ (EMA)

My question's thus two-fold:

 1. Are minibatch statistics computed _per immediate minibatch_, or as an _EMA_? 
 2. How are the inference parameters, shared across all timesteps, `gamma` and `beta` computed? Is the computation in (1) simply averaged across all timesteps? (e.g. average `EMA_t` for all `t`)

<hr>

**Existing implementations**: in Keras and TF below, but are all outdated, and am unsure regarding correctness

 - [Keras](https://github.com/udibr/keras/blob/lstmbn/keras/layers/recurrent.py#L863), [TF-A](https://github.com/tam17aki/recurrent-batchnorm-tensorflow/blob/master/BN_LSTMCell.py), and [TF-B](https://github.com/OlavHN/bnlstm/blob/master/lstm.py)
 - All above agree that during training, _immediate minibatch_ statistics are used, and that `beta` and `gamma` are updated as an EMA of these minibatches
 - Problem: the `bn` operation (in A, and presumably B & C) is applied on a _single timestep_ slice, to be passed to the `K.rnn` control flow for re-iteration. Hence, EMA is computed w.r.t. minibatches _and_ timesteps - which I find questionable:
  - EMA is used in place of a simple average when population statistics are _dynamic_ (e.g. minibatch-to-minibatch), whereas we have access to _all timesteps_ in a minibatch prior having to update `gamma` and `beta`
  - EMA is a worse but at times necessary alternative to a simple average, but per above, we _can_ use latter - so why don't we? Timestep statistics can be cached, averaged at the end, then discarded - holds also for `stateful=True`
